### ISSUE TITLE
* (indicating unsaved work) disappears after clicking save even when cancelled

### ISSUE THREAD

**Issue by dave-doty** [UC-Davis-molecular-computing/scadnano#292]
Open a design. Make a modification. The filename in the tab will have a * to indicate unsaved work.

When saving, this * is removed based on `state.ui_state.changed_since_last_save`. However, it is removed even if the user cancels the save after the Save Dialog window has opened.

Probably `state.ui_state.changed_since_last_save` is being prematurely updated. This should change so that it is updated only upon a successful save.

**Comment by dave-doty** [comment]
**Update:** This seems to be a pretty tough problem to solve, since it involves detecting whether the user hit the cancel button on the Save dialog, but browser security restrictions make it hard to see what the user did with the dialog. (Same reason it is hard to see what new filename they chose, as in #282.

https://github.com/jimmywarting/StreamSaver.js
https://github.com/eligrey/FileSaver.js

Seems to be not possible with FileSaver.js, which appears to use the same approach we are using here.
https://github.com/eligrey/FileSaver.js/issues/126
https://github.com/eligrey/FileSaver.js/issues/187
https://github.com/eligrey/FileSaver.js/issues/389
https://github.com/eligrey/FileSaver.js/issues/569

Seems like it might be possible with StreamSaver.js.

Seems this is a fairly universal problem:
https://stackoverflow.com/questions/41334881/detect-when-user-accepts-to-download-a-file

**Comment by dave-doty** [comment]
I don't think we need to know what happened with the Save dialog. Since we use the nonemptiness of the Undo stack as a proxy for unsaved work, perhaps we can just have middleware that checks the stack and changes the title appropriately when it goes from empty to nonempty and vice versa.

**Comment by dave-doty** [comment]
I removed "high priority" from this, because I realized that it doesn't actually affect whether the user is warned about leaving with unsaved work. It only affects whether the title accurately reflects whether there is unsaved work.

**Comment by UnHumbleBen** [comment]
Are you sure we can use the nonemptiness of the Undo stack? Saving a design doesn't seem to clear the undo stack. I think there are currently false positives in regards to the window warning the user about unsaved work. Ex: If you make an edit and save and then try to close the window, the warning will still pop up, even though the work has been saved.

I guess we could fix both the false-positive warnings and this title issue by making it so that the undo stack changes to empty whenever user saves, but I can easily see this leading to unwanted behavior. For instance, after saving a design, the user might want to undo some steps, but if the undo stack is cleared, then they cannot do that.

We might need a different way to detect unsaved changes besides using undo stack

**Comment by dave-doty** [comment]
I'm sure there is a "standard" logic to handle this. Maybe try searching around a bit and see what people recommend.
